### ISSUE TITLE
Bounding Boxes in Object Recognition are incorrect

### ISSUE THREAD

**Issue by bruden01** [Vantiq/vantiq-extension-sources#103]
The bounding boxes in the object recognition source are incorrect size.  Generally, they are much too big as rendered in the image. 

This is a problem for the ford POC as the partner creating the 3d digital twin needs to use these coordinates to render the 3d twin in the correct location.  At current its unusable by the partner Prefix which is holding up the project.

This is odd though since the coordinates are smaller and in the wrong place compared to runnign it directly through darkflow.

For example for a given image we report
```
"results": [
      {
         "confidence": 0.8291698,
         "location": {
            "top": 48.742752,
            "left": -2.42628,
            "bottom": 448.82892,
            "right": 477.1719
         },
         "label": "person"
```

and darkflow reports
```
[
    {
        "confidence": 0.83,
        "bottomright": {
            "x": 1168,
            "y": 719
        },
        "label": "person",
        "topleft": {
            "x": 284,
            "y": 118
        }
    }
]
```

**Comment by fhcarter** [comment]
Brett,

Part of the issue is that the values are for different images.  Yolo reports its boxes based on a 416x416 image (that's what it's tuned for).  We re-scale these for the actual image size, so it's likely ours will be larger (since the actual images are usually larger than 416x416.

That said, there's clearly something amiss as those values should never be negative (obviously).  We're working on a fix...

**Comment by namirfa** [comment]
@bruden01 

I have merged a partial fix to this issue, which now returns the correctly scaled location values as part of the JSON Object.

Could you try running your image through both the Ext. Source and darkflow again, so we can see if the locations are were we would expect them to be for the differently-sized images. This should help us completely resolve the issue.

**Comment by rblumer4** [comment]
Yes, Namir updated master on 2018-12-21, so you need to rebuild your _vantiq-extension-source.jar_ from the latest master code.

**Comment by bruden01** [comment]
yes,  thats the one I grabbed.  I did a git pull and it says  

```
git pull
Already up to date.
```

i had updated it on the 22nd so I pulled in the change.  But I still don't see the correct results.  I will see there is some chnage as now I saw a number higher than 200.   e.g `"right": 1402.1096`  but the bounding box is not correct. I also verified this with a pb file I created that only does face recogniton.  The bounding box is not correct.  See for comparison.

The outer box is ours and the inner box when I run directly in darkflow.

![image](https://user-images.githubusercontent.com/2548143/50450927-f7958f00-08fe-11e9-9ba7-92929055670a.png)

**Comment by fhcarter** [comment]
Preliminary Results:  The symptoms disappear based upon the model used.  The model that the system "ship" with can recognize 80 different objects, but appears to draw accurate but _less precise_ bounding boxes.  When I change the system to use the ```yolo-voc dataset```, I get more precise results.  However, the VOC dataset (on which that's trained) detects only 20 different objects.

As it turns out, it appears that the code is somewhat tuned to the ```yolo-voc``` data set.  The code, after running the predictor, it figures out what the appropriate bounding boxes are (using some constant modifiers).  We're going to try to figure out how to either derive these from the model _or_ allow for config-based override.

There will be a follow up soon that shows how to use that model (most of which you probably already know since it appears you are using your own model for faces).

**Comment by bruden01** [comment]
That doesn't seem to make sense.  The bounding boxes are correct when I use just the pb file.

The classifier that detects 80 objects is the COCO dataset
The classifier that detect 20 objects is the PASCAL VOC Data Set

The image above is from a PB file I create that's not from either of those that has a single class `face`.  its not COCO or PASCAL.  and the bouding box is correct. 

One thing i have notices when running the PB file directly in darkflow is it requires you to also specify the .meta file that is generated along side the pb file.  for example a meta file looks like

This the meta file for my face recognition PB file
``
{"anchors": [0.738768, 0.874946, 2.42204, 2.65704, 4.30971, 7.04493, 10.246, 4.59428, 12.6868, 11.8741], "classes": 1, "type": "[region]", "coords": 4, "net": {"policy": "steps", "type": "[net]", "steps": "-1,100,80000,100000", "subdivisions": 1, "saturation": 1.5, "momentum": 0.9, "batch": 1, "height": 416, "decay": 0.0005, "angle": 0, "exposure": 1.5, "channels": 3, "width": 416, "hue": 0.1, "max_batches": 120000, "learning_rate": 0.001, "scales": ".1,10,.1,.1"}, "coord_scale": 1, "thresh": 0.6, "jitter": 0.2, "out_size": [13, 13, 30], "labels": ["face"], "num": 5, "bias_match": 1, "softmax": 1, "name": "tiny-yolo-azface-fddb", "absolute": 1, "rescore": 1, "random": 1, "class_scale": 1, "object_scale": 5, "noobject_scale": 1, "model": "/mnt/mbp/YOLO/YOLOV2Face/tiny-yolo-azface-fddb.cfg", "colors": [[254.0, 254.0, 254]], "inp_size": [416, 416, 3]}
``

why doesn't our implentation require this.  Or is it completly unrealted.

When I run against the PB file in darkflow I have to execute like

`flow --pbLoad built_graph/yolo.pb --metaLoad built_graph/yolo.meta --imgdir sample_img/`

**Comment by fhcarter** [comment]
@bruden01 yes,, I think you're correct.  Our's doesn't claim to require it -- but probably should.

The issue you are seeing (that is, what I currently think is the _root cause_ is that our code as the anchors as a constant rather than reading it from somewhere (_i.e._ the .meta file).  The values in the code are ```{1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52}```, which are about a 3rd larger than yours.  That, I think, explains the size differential -- these _anchors_ are used to generate the bounding boxes.

If I look at the .cfg file (which is partially used to generate the .meta file) for a COCO-based file (which is what the default protobuf file in the extension source is), I see that it has anchor values of 
```{0.57273, 0.677385, 1.87446, 2.06253, 3.33843, 5.47434, 7.88282, 3.52778, 9.77052, 9.16828}```, some of which are roughly half of what's in the file -- again approximately the discrepancy.  When I change our code to use those anchors, I get more or less what's expected.

For a validation to try with the supplied .pb file, try the fhc-newanchors branch.  It simply changes the constant value to something closer to accurate.  A more complete fix should be around soon that will be able to read (I think) the meta file and/or allow a direct override in the config file (for special cases...)

**Comment by namirfa** [comment]
@bruden01 
Just sent you an email regarding the recent fix. We have changed the default anchor values so that they are tailored to the `.pb` model that ships with the object recognition source. We have also added an "anchors" option in the Source Configuration that allows the user to override the default anchors, if they are using a different model.

I am closing this issue now.
